### DISCUSSION TITLE
How to enable PictureInPicture for android activities?

### DISCUSSION THREAD

**Discussion by danimarin** [appcelerator/titanium_mobile#12780] [Q&A] [answered]
I am adding 

```
android:resizeableActivity="true" 
android:supportsPictureInPicture="true" 
```

in the main activy at manifest, but when I try to open the PictureInPicture method, I always receive:
 
`Current activity does not support picture-in-picture`
 
But if I create a class in a native module, and launch from native java creating an intent, it works fine. I would like to use it with a titanium window. 

Somehow it seems as the activity properties are lost when opening a new window.

Thanks

**Comment by jquick-axway** [comment, marked as answer]
The main activity is the splash screen window you see on startup.

If you're trying to make a Titanium `Ti.UI.Window` a picture-in-picture window, then you need to apply this attribute to the `org.appcelerator.titanium.TiActivity` instead. You can do so in the "tiapp.xml" as shown below. (Note that all activities are resizable by default so you don't need to add that attribute.)
```xml
<ti:app>
	<android xmlns:android="http://schemas.android.com/apk/res/android">
		<manifest>
			<application>
				<activity android:name="org.appcelerator.titanium.TiActivity" android:supportsPictureInPicture="true"/>
			</application>
		</manifest>
	</android>
</ti:app>
```

This means you would need to call the Java `enterPictureInPictureMode()` method on the `TiActivity` instance which you can get via the window's [activity](https://docs.appcelerator.com/platform/latest/#!/api/Titanium.UI.Window-property-activity) property. You can probably do this via hyperloop or passing this activity instance to your module to call that Java method.

Note that I've never tried pictured-in-picture mode before. Titanium doesn't officially support it, but theoretically the above should make it work.

**Reply by Max87ZA** [reply]
Hi Joshua, can you please add some example code for this? I'm trying to test it, but there is a difference between Java activity and Ti.UI.Window's activity. I was able to call activity.enterBackgroundMode() via hyperloop, but it throwed error:
Error: Activity must be resumed to enter picture-in-picture  So I tried this code: 
`var windowActivity = $.indexWindow.activity;
        var pipActivity = Activity.cast(windowActivity);
        pipActivity.enterPictureInPictureMode(); ` 
but it throw error, that enterPictureInPictureMode does not exists... Please help

**Comment by jquick-axway** [comment]
I tested this out today. I can make picture-in-picture work with a Titanium window in hyperloop, but it'll end up having navigation problems because the system will disconnect the Titanium window/activity from its parent. This completely breaks back navigation.

Unfortunately, this is not something I think we can fix or work-around. From looking at Google's docs and stackoverflow, you're expected to make the picture-in-picture window its own standalone singleTask activity running on a separate task (ie: a separate UI stack). So, I think picture-in-picture activity is best implemented natively... but you can launch it from your Titanium app.

The below is what I've tested with.

```xml
<ti:app>
	<android xmlns:android="http://schemas.android.com/apk/res/android">
		<manifest>
			<application>
				<activity
					android:name="org.appcelerator.titanium.TiActivity"
					android:supportsPictureInPicture="true"
					android:excludeFromRecents="true"
					android:noHistory="true"
					android:autoRemoveFromRecents="true"
					android:allowTaskReparenting="false"/>
			</application>
		</manifest>
	</android>
	<modules>
		<module>hyperloop</module>
	</modules>
</ti:app>
```

```javascript
// app.js
const Activity = require('android.app.Activity');
const parentWindow = Ti.UI.createWindow();
const button = Ti.UI.createButton({ title: "Show Picture-in-Picture" });
button.addEventListener("click", () => {
	if (Ti.Platform.Android.API_LEVEL < 24) {
		alert("Picture-in-Picture is only supported on Android 7 and higher.")
		return;
	}
	const childWindow = Ti.UI.createWindow();
	childWindow.add(Ti.UI.createLabel({ text: "My PiP Window" }));
	childWindow.addEventListener("open", () => {
		const currentActivity = new Activity(childWindow.activity);
		currentActivity.enterPictureInPictureMode();
	});
	childWindow.open();
});
parentWindow.add(button);
parentWindow.open();
```

**Reply by Max87ZA** [reply]
Thanks @jquick-axway your example code works for me. It would be very helpful, that when your window enters into PiP mode and user then clicks on one of two buttons - maximize/close, activity will react to activity.onStop/activity.onResume callbacks(not window callbacks).
So, in your code:
`childWindow.addEventListener("open", () => {
                 childWindow.activity.onStop = function()
{
       //this happens when user clicks on PiP's X button
}
childWindow.activity.onResume = function()
{
       //this happens when user clicks on PiP's maximize button
}
		const currentActivity = new Activity(childWindow.activity);
		currentActivity.enterPictureInPictureMode();
	});
	childWindow.open();`

**Comment by danimarin** [comment]
Thanks everybody, it's working cool.
